File: platform/src/main/java/net/firejack/platform/core/model/registry/site/NavigationElementModel.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package net.firejack.platform.core.model.registry.site;

import net.firejack.platform.api.registry.model.RegistryNodeStatus;
import net.firejack.platform.core.model.registry.*;
import net.firejack.platform.core.model.registry.authority.PermissionModel;
import org.hibernate.annotations.ForeignKey;

import javax.persistence.*;
import java.util.List;


@javax.persistence.Entity
@DiscriminatorValue("NAV")
public class NavigationElementModel extends RegistryNodeModel implements ISortable, INavigableRegistryNode, IAllowDrag, IAllowDrop, IFolder, IAllowCreateAutoDescription {

    private static final long serialVersionUID = -1209284431447149479L;
    private String serverName;
    private String parentPath;
    private String urlPath;
    private String urlParams;
    private Integer port;
    private EntityProtocol protocol;
    private RegistryNodeStatus status;
    private String pageUrl;
    private Boolean hidden;
    private NavigationElementType elementType;

    private List<PermissionModel> permissions;

    @Column(length = 1024)
    public String getServerName() {
        return serverName;
    }

    public void setServerName(String serverName) {
        this.serverName = serverName;
    }

    public Integer getPort() {
        return port;
    }

    public void setPort(Integer port) {
        this.port = port;
    }

    @Column(length = 255)
    public String getParentPath() {
        return parentPath;
    }

    public void setParentPath(String parentPath) {
        this.parentPath = parentPath;
    }

    @Column(length = 2048)
    public String getUrlPath() {
        return urlPath;
    }

    public void setUrlPath(String urlPath) {
        this.urlPath = urlPath;
    }

    @Column(length = 2048)
    public String getUrlParams() {
        return urlParams;
    }

    public void setUrlParams(String urlParams) {
        this.urlParams = urlParams;
    }

    /**
     * @return
     */
    @Enumerated
    public EntityProtocol getProtocol() {
        return protocol;
    }

    /**
     * @param protocol
     */
    public void setProtocol(EntityProtocol protocol) {
        this.protocol = protocol;
    }

    /**
     * @return
     */
    @Enumerated
    public RegistryNodeStatus getStatus() {
        return status;
    }

    /**
     * @param status
     */
    public void setStatus(RegistryNodeStatus status) {
        this.status = status;
    }

    @ManyToMany(targetEntity = PermissionModel.class, fetch = FetchType.LAZY)
    @JoinTable(
            name = "opf_navigation_element_permission",
            joinColumns = @JoinColumn(name = "id_navigation_element", referencedColumnName = "ID"),
            inverseJoinColumns = @JoinColumn(name = "id_permission", referencedColumnName = "ID")
    )
    @ForeignKey(name = "FK_NAV_ELEMENT_PERMISSIONS")
    public List<PermissionModel> getPermissions() {
        return permissions;
    }

    public void setPermissions(List<PermissionModel> permissions) {
        this.permissions = permissions;
    }

    @Column(length = 2048)
    public String getPageUrl() {
        return pageUrl;
    }

    public void setPageUrl(String pageUrl) {
        this.pageUrl = pageUrl;
    }

    @Column(name = "hidden")
    public Boolean getHidden() {
        return hidden;
    }

    public void setHidden(Boolean hidden) {
        this.hidden = hidden;
    }

    @Enumerated
    @Column(name = "element_type")
    public NavigationElementType getElementType() {
        return elementType;
    }

    public void setElementType(NavigationElementType elementType) {
        this.elementType = elementType;
    }

    @Override
    @Transient
    public RegistryNodeType getType() {
        return RegistryNodeType.NAVIGATION_ELEMENT;
    }

}

File: platform/src/main/webapp/js/net/firejack/platform/console/site/model/NavigationElementModel.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

Ext.define('OPF.console.site.model.NavigationElement', {
    extend: 'Ext.data.Model',

    statics: {
        pageSuffixUrl: 'console/site',
        restSuffixUrl: 'site/navigation',
        editorClassName: 'OPF.console.site.editor.NavigationElementEditor',
        constraintName: 'OPF.site.NavigationElement'
    },

    idProperty: 'id',

    fields: [
        { name: 'id', type: 'int', useNull: true },
        { name: 'name', type: 'string' },
        { name: 'path', type: 'string' },
        { name: 'description', type: 'string' },
        { name: 'lookup', type: 'string' },

        { name: 'parentId', type: 'int' },

        { name:'serverName', type: 'string'},
        { name:'port', type: 'string'},
        { name:'parentPath', type: 'string'},
        { name:'urlPath', type: 'string'},
        { name:'protocol', type: 'string'},
        { name:'method', type: 'string'},
        { name:'status', type: 'string'},
        { name:'sortPosition', type: 'string'},
        { name:'elementType', type: 'string'},

        { name: 'created', type: 'int' },
        { name: 'canUpdate', type: 'boolean' },
        { name: 'canDelete', type: 'boolean' },

        { name: 'navigationElements', type: 'auto' },

        { name: 'permissions', type: 'auto'},

        { name: 'textResourceVersion', type: 'auto'},

        { name: 'imageResourceVersion', type: 'auto'}
    ],

    hasMany: [
        { model: 'NavigationElement', name: 'navigationElements', foreignKey: 'parent_id' }//,
//        { model: 'Permission', name: 'permissions' }
    ]

});